ntoskrnl/io/iomgr/deviface.c

```diff
@@ -30,185 +31,450 @@ static PWCHAR BaseKeyString = L"\\Registry\\Machine\\System\\CurrentControlSet\\
+        {
+            break;
+        }
     }
-    GuidString.Buffer = StartPosition;
-    GuidString.MaximumLength = GuidString.Length = (USHORT)((ULONG_PTR)(EndPosition + 1) - (ULONG_PTR)StartPosition);
+    ReferenceStringOffset = i * sizeof(WCHAR);
 
-    InitializeObjectAttributes(&ObjectAttributes,
-                               &GuidString,
-                               OBJ_CASE_INSENSITIVE | OBJ_KERNEL_HANDLE,
-                               ClassesKey,
-                               NULL);
-    Status = ZwCreateKey(GuidKeyRealP,
-                         DesiredAccess | KEY_ENUMERATE_SUB_KEYS,
-                         &ObjectAttributes,
-                         0,
-                         NULL,
-                         REG_OPTION_NON_VOLATILE,
-                         NULL);
-    ZwClose(ClassesKey);
+    /* The GUID is before the reference string or at the end */
+    ASSERT(SymbolicLinkName->Length >= ReferenceStringOffset);
+    if (ReferenceStringOffset < GUID_STRING_BYTES + sizeof(WCHAR))
+    {
+        DPRINT1("Invalid link name '%wZ'\n", SymbolicLinkName);
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    /* Get reference string from link without prefix */
+    // TODO: Add some more sanity checks
+    ReferenceStringReal.Buffer = LinkNameNoPrefix.Buffer + ReferenceStringOffset / sizeof(WCHAR);
+    ReferenceStringReal.Length = LinkNameNoPrefix.Length - ReferenceStringOffset;
+    ReferenceStringReal.MaximumLength = ReferenceStringReal.Length;
+
+    /* Get GUID string */
+    GuidStringReal.Buffer = LinkNameNoPrefix.Buffer + (ReferenceStringOffset - GUID_STRING_BYTES) / sizeof(WCHAR);
+    GuidStringReal.Length = GUID_STRING_BYTES;
+    GuidStringReal.MaximumLength = GuidStringReal.Length;
+
+    /* Allocate a temporary device string (just PCI and Vendor IDs without prefix, GUID and reference strings) */
+    DeviceString.MaximumLength = LinkNameNoPrefix.Length - ReferenceStringReal.Length;
+    DeviceString.Length = DeviceString.MaximumLength;
+    DeviceString.Buffer = ExAllocatePoolWithTag(NonPagedPool, DeviceString.MaximumLength, TAG_IO);
+    if (!DeviceString.Buffer)
+    {
+        DPRINT1("ExAllocatePoolWithTag() failed\n");
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    /* Fill it by an appropriate part of the link without prefix */
+    RtlCopyUnicodeString(&DeviceString, &LinkNameNoPrefix);
+
+    /* Allocate a munged path string */
+    MungedStringReal.Length = 0;
+    MungedStringReal.MaximumLength = MungedPrefix.Length + DeviceString.Length;
+    MungedStringReal.Buffer = ExAllocatePoolWithTag(NonPagedPool, MungedStringReal.MaximumLength, TAG_IO);
+    if (!MungedStringReal.Buffer)
+    {
+        DPRINT1("ExAllocatePoolWithTag() failed\n");
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    /* Fill it by munged prefix and munged string */
+    Status = RtlAppendUnicodeStringToString(&MungedStringReal, &MungedPrefix);
     if (!NT_SUCCESS(Status))
```

Comment:
Like above, this can't fail so just assert.

Reply:
Done.